mode_calculations.py

```diff
@@ -600,3 +601,422 @@ def inner_product(t, abar, b, axis=None, apply_conjugate=False):
+    Because of the magnetic selection rules, we only have mp == m
+
+    We could have used `_swsh_Y_mat_el` but I am just preemptively
+    combining the prefactors.
+    """
+
+    s = -2
+
+    for ell in range(ell_min, ell_max+1):
+        ellp_min = max(ell_min, ell - 1)
+        ellp_max = min(ell_max, ell + 1)
+        for ellp in range(ellp_min, ellp_max+1):
+            for m in range(-ell, ell+1):
+                if ((m < -ellp) or (m > ellp)):
+                    continue
+                cg1 = CG(ell, m,  1, 0, ellp, m)
+                cg2 = CG(ell, -s, 1, 0, ellp, -s)
+                prefac = np.sqrt( (2.*ell + 1.) / (2.*ellp + 1.) )
+                yield ellp, m, ell, m, (prefac * cg1 * cg2)
+
+def _make_p_plus_minus_for_s_minus_2(sign):
+    u"""Produce the function p_plus_M_for_s_minus_2 or p_minus_M_for_s_minus_2, based on sign.
+
+    This function is specific to the case where waveforms have s=-2.
+
+    p^+ = -\sqrt{8 \pi / 3} Y_{1,+1} = \sin\theta e^{+i\phi}
+    p^- = +\sqrt{8 \pi / 3} Y_{1,-1} = \sin\theta e^{-i\phi}
+    This is what Ruiz+ (2008) [0707.4654] calls "l^±", which is a bad name.
+
+    We use `_swsh_Y_mat_el` to compute the matrix elements.
+    Notice that since the operator has definite m = ±1, we only have
+    mp == m ± 1 nonvanishing in the matrix elements.
+
+    TODO maybe make a docstring for the emitted function?
+    """
+
+    sign = float(sign)
+
+    if (sign != 1.) and (sign != -1.):
+        raise ValueError("sign must be either 1. or -1. in make_j_plus_minus")
+
+    s = -2
+    prefac = -1. * sign * np.sqrt( 8. * np.pi / 3. )
+
+    def p_pm_M(ell_min, ell_max):
+        for ell in range(ell_min, ell_max+1):
+            ellp_min = max(ell_min, ell - 1)
+            ellp_max = min(ell_max, ell + 1)
+            for ellp in range(ellp_min, ellp_max+1):
+                for m in range(-ell, ell+1):
+                    mp = round(m + 1 * sign)
+                    if ((mp < -ellp) or (mp > ellp)):
+                        continue
+                    yield (ellp, mp, ell, m,
+                           (prefac *
+                            _swsh_Y_mat_el(s, ellp, mp, 1., sign, ell, m)))
+
+    return p_pm_M
+
+p_plus_M_for_s_minus_2  = _make_p_plus_minus_for_s_minus_2(1.)
+p_minus_M_for_s_minus_2 = _make_p_plus_minus_for_s_minus_2(-1.)
+
+def energy_flux(h):
+    """Compute energy flux from waveform
+
+    This implements Eq. (2.8) from Ruiz+ (2008) [0707.4654].
+    """
+    from .waveform_modes import WaveformModes
+    from . import h as htype
+    from . import hdot as hdottype
+    if not isinstance(h, WaveformModes):
+        raise ValueError("Momentum flux can only be calculated from a `WaveformModes` object; "
+                         +"this object is of type `{0}`.".format(type(h)))
+    if h.dataType == hdottype:
+        hdot = h
+    elif h.dataType == htype:
+        hdot = h.copy()
+        hdot.dataType = hdottype
+        hdot.data = h.data_dot
```

Comment:
Since you are just passing in arrays as arguments to `np.einsum` on line 911, I don't think making the new WaveformModes objects here are necessary. You could just do something like:
```
if h.dataType = hdottype:
    hdot = h.data
elif h.dataType == htype:
    hdot = h.data_dot
```
Then in your `np.einsum` command, you just have:
```
Edot = np.einsum('ij, ij -> i', hdot.conjugate(), hdot).real
```

Reply:
Good point! Done, will push a commit in a min.